Synthetic Data Generation Using Diffusion Models (#125)
johko/computer-vision-course: chapters/en/Unit 10 - Synthetic Data Creation/datagen-diffusion-models.mdx

# Synthetic Data Generation with Diffusion Models

> Imagine trying to train a model for tumor segmentation, as it's hard to gather data for medical imaging, it'd be really difficult for the model to converge.

Reviewer: ```suggestion
Imagine trying to train a model for tumor segmentation. As it's hard to gather data for medical imaging, it'd be really difficult for the model to converge.
```

Ideally, we expect to have at least enough data to build a simple baseline, but what if you have just a few samples?

> Synthetic data generation methods try to solve this dilemma, and now we have many more options with the boom of generative models!

Reviewer: You can continue this sentence from previous one not to break the flow

> As you've seen in the previous sections, it is possible to use generative models such as DCGAN to generate synthetic images. In this section, we will focus on diffusion models using `diffusers`!

Reviewer: Maybe you can give link to diffusers library

> ## Remember the Diffusion Models

Reviewer: ```suggestion
## Diffusion Models
```

> Diffusion models are generative models that gained significant popularity in recent years, thanks to their capabilities to produce high-quality images. They're widely used in image, video and text synthesis nowadays.

Reviewer: ```suggestion
Diffusion models are generative models that gained significant popularity in recent years, thanks to their capabilities to produce high-quality images. Nowadays, they're widely used in image, video, and text synthesis.
```

> The diffusion model works by learning to denoise random Gaussian noise step-by-step, the training process requires adding Gaussian noise to the input samples and letting the model learn denoising.

Reviewer: ```suggestion
A diffusion model works by learning to denoise random Gaussian noise step-by-step. The training process requires adding Gaussian noise to the input samples and letting the model learn denoising.
```

> These models are generally conditioned to some kind of input besides the data distribution, such as text prompts, images or even audio. But it's also possible to [build an unconditional generator](https://huggingface.co/docs/diffusers/training/unconditional_training).

Reviewer: I think the newlines are breaking the flow in readability

While there are many underlying concepts behind the model's inner workings, the simplified version is similar to this:

> The model adds noise to the input and processes it

Reviewer: ```suggestion
Firstly, the model adds noise to the input and processes it.
```

> ![noising](https://i.imgur.com/dR7E3D1.jpeg)

Reviewer: You can add the images to hf-vision/course-assets repository, so that the links will never break

> Then the model learns to denoise the given data distribution

Reviewer: ```suggestion
Then the model learns to denoise the given data distribution.
```

![denoising](https://i.imgur.com/Lc13i0S.jpeg)

We won't dive deep into the theory, but understanding how a diffusion model works will come really handy when we need to pick a technique.

## Text-To-Image Diffusion Model: Stable Diffusion

> Essentially, the way SD works is the same as we mentioned above. It uses three main components that help it to produce high-quality images.

Reviewer 2: Little suggestion, it'd be better to spell out the acronym on the first use and follow with the abbreviation in parentheses to avoid confusion. For example, use Stable Diffusion (SD) the first time, then use SD freely in subsequent mentions.

Author: I thought the usage in the title would count as the first usage but I agree, it looks nicer that way. Thanks for your feedback!

> Essentially, the way SD works is the same as we mentioned above. It uses three main components that help it to produce high-quality images.

Reviewer: ```suggestion
Essentially, the way Stable Diffusion works is the same as we mentioned above. It uses three main components that help it to produce high-quality images.
```
Maybe you could define the Stable diffusion at first

> 1. __The diffusion process:__ The input is processed multiple times to generate useful information about the image. The "usefulness" is learned while training the model.

Reviewer: ```suggestion
1. **The diffusion process:** The input is processed multiple times to generate useful information about the image. The "usefulness" is learned while training the model.
```
What do you mean with usefulness? Semantic info?

> 2. __Image encoder and decoder model:__ It lets the model compress images from pixel space to a smaller dimensional space, abstracting meaningless information while increasing performance.

Reviewer: ``` suggestion
2. **Image encoder and decoder model:** Lets the model compress images from pixel space to a smaller dimensional space, abstracting meaningless information while increasing performance.
```

> 3. __Optional conditional encoder:__ This component is used to condition to generation process on an input. This extra input can be text prompts, images, audio and other kinds of representations. Originally it was a text encoder

Reviewer: ```suggestion
3. **Optional conditional encoder:** This component is used to condition to generation process on an input. This extra input can be text prompts, images, audio, and other representations. Originally, it was a text encoder.
```

So while the general workflow looks like this:

> 

Reviewer: ```suggestion
```

![general stable diffusion workflow](https://i.imgur.com/9ZcNijf.png)

Originally we used a text encoder to condition the model on our prompts:

> 

Reviewer: ```suggestion
```

![original stable diffusion workflow](https://i.imgur.com/FVeHRff.png)

> This short overview is just the tip of the iceberg! If you wish to dive deep into the theory behind stable diffusion (or diffusion models), you can check these out:

Reviewer: You can put these references in the end not to break the readability

| Text-Guided Image Generation | generate an image given a text prompt |[conditional_image_generation](https://huggingface.co/docs/diffusers/using-diffusers/conditional_image_generation) |
| Text-Guided Image-to-Image Translation | adapt an image guided by a text prompt | [img2img](https://huggingface.co/docs/diffusers/using-diffusers/img2img) |
| Text-Guided Image-Inpainting | fill the masked part of an image given the image, the mask and a text prompt | [inpaint](https://huggingface.co/docs/diffusers/using-diffusers/inpaint) |
| Text-Guided Depth-to-Image Translation | adapt parts of an image guided by a text prompt while preserving structure via depth estimation | [depth2img](https://huggingface.co/docs/diffusers/using-diffusers/depth2img) |

There's also a complete list of supported tasks that you can find from the [Diffusers Summary](https://huggingface.co/docs/diffusers/api/pipelines/overview#diffusers-summary) table.

This means we have many tools under our belt to generate synthetic data!

## Approaches to Synthetic Data Generation

There are generally three cases for needing synthetic data,

> 

Reviewer: ```suggestion
```

> _Extending an existing dataset_:

Reviewer: ```suggestion
**Extending an existing dataset:**
```

- There are not enough samples.

>   - A nice example is a medical imaging dataset such as [DDSM](https://www.mammoimage.org/databases/) (Digital Database for Screening Mammography, ~2500 samples), a small amount of samples makes it harder to build a model for further analysis. It's also rather expensive to build such medical imaging datasets.

Reviewer: if there's a single sub-element you can do: 
- **There are not enough samples:** A nice example is a medical imaging dataset such as ..

_Creating a Dataset from Scratch_:

- There aren't any samples at all.
  - Let's assume that you want to build a weapon detection system on top of CCTV video streams. But there aren't any samples for the specific weapon you want to detect, you can use similar observations in different settings to apply style transfer to make them look like CCTV streams!

_Preserving Privacy_:

- Hospitals collect huge amounts of data on patients, surveillance cameras capture raw information about individuals' faces and activities, and all of these introduce a potential infringe on privacy. We can use diffusion models to generate privacy-preserving datasets to develop our solutions, without giving up on anyone's privacy rights.

There are different methods for us to utilize a text-to-image diffusion model to generate customized outputs:

> _Just prompt engineering_:

Reviewer: ```suggestion
**Prompt Engineering:**
```

> By simply utilizing the pre-trained diffusion model (such as [Stable Diffusion XL](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/stable_diffusion_xl)), you can try to construct a nice prompt to generate images. But the quality of the generated images may not be consistent and it might be really difficult to construct such a prompt for your specific use case.

Reviewer: You can continue this line from above line for readability

_Training with [Textual Inversion](https://huggingface.co/docs/diffusers/main/en/training/text_inversion)_:

![textual-inversion](https://i.imgur.com/d8ZnaZD.png)

Textual inversion is a technique that works by interfering with the text embeddings in the model's architecture. You can add a new token to the vocabulary and then fine-tune the embeddings using a few examples.

By providing samples corresponding to the new token, we try to optimize the embeddings to capture the characteristics of the object.

_Training a [LoRA (Low-Rank Adaptation)](https://huggingface.co/docs/diffusers/main/en/training/lora?installation=PyTorch) model_:

> LoRA has targeted the problem of fine-tuning LLMs. It's a technique that injects new weights into the pre-trained model and tries to optimize those.

Reviewer 2: I think this line doesn't convey the real significance of LoRA. It's not just injecting weights. Wouldn't it be nice to indicate that: In LoRA, we freeze all model parameters and train a lower-rank subset (i.e. incredibly fewer parameters), which are then merged with the frozen weights introducing no inference latency?

Author: I thought it'd introduce a little too much complexity to the technique but you're right, LoRA is essentially a nice technique because of training lower-rank update matrices. I think updating the LoRA part as the following would be nice:

> LoRA has targeted the problem of fine-tuning LLMs. It represents the weight updates with two smaller update matrices through low-rank decomposition, which has substantially less amount of parameters. Then these matrices can be trained to adapt the new data!
> 
> The base model's weights remain frozen in the whole process, so we just train the new update matrices. Then lastly, the adapted weights are combined with the original weights.
> 
> This means training LoRA is essentially much faster than the model itself! As an important note, it's also possible to combine LoRA with other techniques since it can be added on top of the model itself.

What do you think?

Reviewer 2: Amazing! That sounds great! 
> This means training LoRA is essentially much faster than the model itself!
Just a small suggestion "much faster than full model fine-tuning!"

This means training LoRA is essentially much faster than the model itself! As an important note, it's also possible to combine LoRA with other techniques since it can be added on top of the model itself.

_Training with [DreamBooth](https://huggingface.co/docs/diffusers/main/en/training/dreambooth)_:

![dreambooth](https://i.imgur.com/yoh3RvQ.png)

> DreamBooth is changing the diffusion model completely. Given a few images of a subject, it lets you fine-tune a pre-trained text-to-image model. The main idea is to associate a unique identifier with that specific subject.

Reviewer: I don't think it changes the model completely, rather it's a fine-tuning technique

Author: In the original paper, the authors mentioned the best results were achieved by fine-tuning all the layers of the model. So I wrote it like that since the [explanation on the diffusers documentation](https://huggingface.co/docs/diffusers/main/en/training/dreambooth) also stated that. But something like the following may clear the potential confusion because of the wording:

>DreamBooth is a technique to fine-tune the model to personalize the outputs. Given a few images of a subject, it lets you fine-tune a pre-trained text-to-image model. The main idea is to associate a unique identifier with that specific subject.
>
>For training, we use the tokens in the vocabulary and build a dataset using preferably a rare-token identifier. Because if you choose a rather common identifier, the model would also have to learn to disentangle from their original meaning.
>
>In the original paper, authors find the rare tokens in the vocabulary and then choose identifiers from those. This reduces the risk of an identifier having a strong prior. It is also stated that the best results were achieved by fine-tuning all the layers of the model.

For training, we use the tokens in the vocabulary and build a dataset using preferably a rare-token identifier. Because if you choose a rather common identifier, the model would also have to learn to disentangle from their original meaning.

In the original paper, authors find the rare tokens in the vocabulary and then choose identifiers from those. This reduces the risk of an identifier having a strong prior.

_Training with [Custom Diffusion](https://huggingface.co/docs/diffusers/main/en/training/custom_diffusion)_:

![custom diffusion](https://i.imgur.com/427EPSs.png)

Custom Diffusion is a really powerful technique to personalize the model. It requires just a few samples like the previously mentioned methods, but its power lies in being able to learn multiple concepts at the same time!

It works by training just a part of the diffusion process and the text encoder we mentioned above, meaning that there are fewer parameters to optimize. Thus, this method also enables fast fine-tuning!

## Real-world Examples of Using Diffusion Models for Dataset Generation

There are many unique cases where diffusion models are used to generate synthetic datasets!

> Such as:

Reviewer: ```suggestion
Find them below.
```
